### ISSUE TITLE
Polycyclic 2.16 timeouts when all packages loaded (another regression in AddToIgs?)

### ISSUE THREAD

**Issue by olexandr-konovalov** [gap-packages/polycyclic#46]
This happens both in stable-4.11 and master branches of GAP:
```
Testing polycyclic 2.16, test=/circa/scratch/gap-jenkins/workspace/GAP-master-test/GAPCOPTS/64build/GAPTARGET/packages/label/kovacs/GAP-master-snapshot/pkg/polycyclic-2.16/tst/testall.g, all packages=true
Build timed out (after 900 minutes). Marking the build as aborted.
Build was aborted
```
When GAP started with `-A` or with default packages, tests pass as normally.

If you want to try to reproduce it, start GAP, call `LoadAllPackages();`, and then `TestPackage("polycyclic");`.

**Comment by alex-konovalov** [comment]
See e.g. now a public version of the test: https://travis-ci.org/github/gap-infra/gap-docker-pkg-tests-master/jobs/714790608

**Comment by fingolfin** [comment]
First off, thank you for the report!

Unfortunately the debug log you linked to is almost completely useless, as it does not show the output of the test suite, which normally would tell me where the hang occurs. The problem is that GAP's `make testpackage` apparently first prints all test results into a file, and only afterwards prints the complete log. That's suboptimal. It also disables the progressive log feature, and thus makes the CI test runs less responsive (as in: instead of seeing how the tests proceed, we see nothing for a potentially looong time and then the complete test output at once). This suggests to me that `make testpackage` should be improved.

A local test quickly revealed that it hangs in `tst/bugfix.tst` line 38:
```
...
gap> g := PcGroupToPcpGroup(SmallGroup(48,1));
Pcp-group with orders [ 2, 2, 2, 2, 3 ]
gap> # The next two commands used to trigger errors
gap> NonAbelianTensorSquare(Centre(g));  # <--- this hangs
Pcp-group with orders [ 8 ]
...
```
More specifically, this hangs:
```gap
Centre(PcGroupToPcpGroup(SmallGroup(48,1)))
```

Since this works in a regular session, and only fails when all packages are loaded, I suspect it has to do with method ranking changes, but I am not sure yet.

**Comment by alex-konovalov** [comment]
Thanks for looking at this. Yes, `make testpackage` should be improved in the same way like testinstall etc., - with `tee`.

**Comment by alex-konovalov** [comment]
@fingolfin thanks for your earlier investigation. This still happened (as of yesterday), and causes jenkins tests backlogs and then termination. It would help me to check the status of the package tests before the next GAP release, if we could fix this somehow.

**Comment by fingolfin** [comment]
Let `G:=PcGroupToPcpGroup(SmallGroup(48,1)));`. Then the hang also happens with `FittingSubgroup(G)`, and that in turn with  `PCore(G, 2)`, resp.  `Core( G, SylowSubgroup( G, p ) )`:
```gap
G:=PcGroupToPcpGroup(SmallGroup(48,1)); S:=SylowSubgroup(G,2); Core(G,S);
```
hangs. And *this* hang I can reproduce even in a gap loaded with `-A` and in which I only loaded polycyclic (and its dependencies, of course).

And it involves `AddToIgs`, I am afraid...

**Comment by alex-konovalov** [comment]
Ok, so it works in GAP 4.11.0 with Polycyclic 2.15.1, and timeouts in GAP master with Polycyclic 2.16. After replacing Polycyclic 2.16 by Polycyclic 2.15.1 in the GAP master, it runs fine. So the problem is likely in polycyclic. Two commit messages since 2.15.1 mention `AddToIgs`, although it could be some other commit triggering it...

**Comment by alex-konovalov** [comment]
I should probably disable 2.16 now, rolling back to the stable 2.15.1 in the package archive, and this will help me to check the state of all other packages in the GAP distribution...

**Comment by fingolfin** [comment]
However, if you roll back to 2.15.1, then you'll reintroduce other bugs.

**Comment by fingolfin** [comment]
Now that I have a good way to reproduce the issue, I can try to look into fixing `AddToIgs`... But it'll require some more time.

**Comment by alex-konovalov** [comment]
The break loop when I terminate it points indeed to a location touched by those two new commits changing `AddToIgs`. 

Yes, I realise it will reintroduce other bugs - but at least now I will be able to run Jenkins job with all other package tests in all combinations (-A, default, all packages), see it completed and evaluate. May later decide what to do with 2.16, maybe you will fix by then...

**Comment by fingolfin** [comment]
I would also suggest to revise that test setup to impose a timeout limit on individual package tests; not doing so seems like a flaw in the test setup

**Comment by alex-konovalov** [comment]
@fingolfin yeah... it now runs the whole `make testpackages` in a single job, with a time limit on its whole duration. What is  easy to maintain in jenkins, and easy to run with `make testpackages` without Jenkins. What would be a good cross-platform solution to implement a timeout in a shell script?

**Comment by sTertooy** [comment]
tl;dr: `IsNormal` may incorrectly return `true` for subgroups that do not normalise each other.

---

I think I may have figured out what's happening here. In @fingolfin's example, the operation `CoreOp` will try to calculate the following intersection:
```
gap> G := PcGroupToPcpGroup( SmallGroup( 48, 1 ) );;
gap> S := SylowSubgroup( G, 2 );; 
gap> T := S^G.5;;
gap> Intersection( T, S );
Pcp-group with orders [ 2, 2, 2, 2 ]
```
The result given by `Intersection` is, in fact, wrong, causing `CoreOp` to get stuck in an infinite loop. The problem is that `Intersection` will check if `S` normalises `T` (or vice versa), and will call `NormalIntersection` if so. However, while the groups `S` and `T` do not normalise each other, `IsNormal` seems to think they do:
```
gap> S.1^T.1 in S;
false
gap> T.1^S.1 in T;
false
gap> IsNormal( S, T );
true
gap> IsNormal( T, S );
true
```
The method installed for `IsNormal` is the following:
https://github.com/gap-packages/polycyclic/blob/929755ef354319268a7cd8cf574fad92543f38f6/gap/basic/pcpgrps.gi#L162-L172
Nothing will actually be checked for the groups in the example, because:
```
gap> Pcp( S, T );
Pcp [  ] with orders [  ]
```
Afaik, the two-argument version of `Pcp` is meant to be used with a normal subgroup as second argument, so its use inside `IsNormalOp` seems odd. Replacing it with the one-argument version or simply disabling this method entirely fixes the problem - but `Intersection( S, T )` can no longer be calculated (#23). 

Before the change in df055111a6679129b05951b7bc102c1b804d64cb, `NormalIntersection` did actually return the correct intersection of `S` and `T` in the example above, which is why the code didn't hang before.

With `IsNormal` fixed, we no longer get a hang when all packages are loaded, but we do get 
```
testing: /usr/lib/gap/pkg/polycyclic/tst/bugfix.tst
########> Diff in /usr/lib/gap/pkg/polycyclic/tst/bugfix.tst:38
# Input is:
NonAbelianTensorSquare(Centre(g));
# Expected output:
Pcp-group with orders [ 8 ]
# But found:
Error, sorry: intersection for non-normal groups not yet installed
########
########> Diff in /usr/lib/gap/pkg/polycyclic/tst/bugfix.tst:40
# Input is:
NonAbelianExteriorSquare(Centre(g));
# Expected output:
Pcp-group with orders [  ]
# But found:
Error, sorry: intersection for non-normal groups not yet installed
########
    4545 ms (466 ms GC) and 319MB allocated for bugfix.tst
```
The problem there indeed seems to be method ranking changes for `FittingSubgroup`. Before loading all packages, we have
```
gap> RankFilter( IsGroup and IsFinite );
47
gap> RankFilter( IsPcpGroup );
50
```
and after loading all packages we have
```
gap> RankFilter( IsGroup and IsFinite );
73
gap> RankFilter( IsPcpGroup );
64
```
which causes the generic method for finite groups to be used, rather than the PCP-version.

**Comment by alex-konovalov** [comment]
So much wish for the new release of Polycyclic - the tests of all packages can't be completed due to the timeout... Any hope that it is in some releasable state soon?

**Comment by alex-konovalov** [comment]
There were 12 commits since the last release in summer 2020. Any chances for a new release?

**Comment by jamesjer** [comment]
Due to the test suite hang with gap 4.12.0 and 4.12.1, I added https://github.com/gap-packages/polycyclic/commit/3f385e49fca33917bfcf5f6d61828bdd97c4468a to the Fedora package build.  This does allow the test suite to run to completion.  However, now the aclib test suite fails:

```
testing: /builddir/build/BUILDROOT/gap-pkg-aclib-1.3.2-8.fc38.x86_64/usr/lib/g\
ap/pkg/aclib/tst/manual.example-3.tst
########> Diff in /builddir/build/BUILDROOT/gap-pkg-aclib-1.3.2-8.fc38.x8\
6_64/usr/lib/gap/pkg/aclib/tst/manual.example-3.tst:54
# Input is:
AlmostCrystallographicInfo(G);
# Expected output:
rec( dim := 4, param := [ -3, 2, -1, -2, -1 ], type := 70 )
# But found:
rec( dim := 4, param := [ 1, -4, 0, 1, 0 ], type := 70 )
########
########> Diff in /builddir/build/BUILDROOT/gap-pkg-aclib-1.3.2-8.fc38.x8\
6_64/usr/lib/gap/pkg/aclib/tst/manual.example-3.tst:60
# Input is:
AlmostCrystallographicInfo(G);
# Expected output:
rec( dim := 4, param := [ -1, 1, -4, 1, 0 ], type := 70 )
# But found:
rec( dim := 4, param := [ -2, 3, 1, 0, -2 ], type := 70 )
########
      54 ms (34 ms GC) and 2.74MB allocated for manual.example-3.tst
testing: /builddir/build/BUILDROOT/gap-pkg-aclib-1.3.2-8.fc38.x86_64/usr/lib/g\
ap/pkg/aclib/tst/manual.example-4.tst
     162 ms (31 ms GC) and 22.3MB allocated for manual.example-4.tst
-----------------------------------
total       216 ms (65 ms GC) and 25.1MB allocated
              2 failures in 1 of 2 files

#I  Errors detected while testing
```

I also tried building polycyclic from git head, in case some other patch fixed the aclib issue.  However, there is no change in the aclib test suite failure.  Please advise on what I should do for Fedora 37.  Should I:

1. Ship polycyclic 2.16 with the test suite disabled, possibly causing users to encounter the hang; or
2. Ship polycyclic 2.16 plus the indicated commit (or polycyclic git head) to avoid the hang and ship aclib with its test suite disabled, possibly giving users a broken aclib?

**Comment by fingolfin** [comment]
Option 1 is my clear preference

**Comment by stertooy** [comment]
For what it's worth, aclib's test suite fails because it generates groups with random parameters, but the expected test output has these parameters hardcoded. Presumably the change with FittingSubgroup now changes which random parameters are generated in the test suite... but nothing is wrong with either package itself.
